### ISSUE TITLE
speed up downloads

### ISSUE THREAD

**Issue by hiromipaw** [micahflee/onionshare#1295]
At the last state of the onion we did a demo about the network performances and how you could get more speed
over onions (https://youtu.be/IyWyTypRGWQ?t=2028).

I was wondering if it makes sense for onionshare to ship with a download accelerator mode.

The idea would be to download chunks of the file in parallel, using http range requests. Each http request must have either a separate socks username+password.

What do you think?

**Comment by micahflee** [comment]
I think it's a great idea. It would mean the person downloading files would use OnionShare itself instead of just Tor Browser, so maybe making a download accelerator tab type or something. Also, it would mean supporting http range requests, which [we never finished making work](https://github.com/micahflee/onionshare/pull/804).

**Comment by mikeperry-tor** [comment]
Awesome! Not to talk too much out of school, but this is a requested feature from journalists who sometimes receive very large files from sources over onion services. There is much interest in this feature. If you ask around, perhaps you will get people interested in helping. This might be a good Tor Hack Week project (March 29-April 2).

Some points of guidance:

1. The limiting factor that makes Tor slow now is the fixed-size windows of 500 cells for streams, and 1000 cells for circuits. This caps downloads at ~500k/sec for exit downloads (depending on latency), and half that for onion services (because paths are twice as long).
2. During a stress test that removed these limits on a custom client and custom service, downloads over 6-hop onion service circuits hit 6MB/sec. This was demonstrated lived in the performance demo.
3. Our performance work will remove these limitations, but likely won't be available for clients for at least 9 months. Additionally, traffic splitting across multiple circuits at the tor layer, which we are also investigating, will take even longer, and will only use 2 circuits. 
4. Both of these features are of the nature that only clients and onion services need to upgrade to benefit. The network does not need to change for this use case (only for exit downloads).
5. During the performance demo, we noticed that even headless onionshare services could not hit the 6MB limit, but lighthttpd/apache/nginx backed services could. Something additional may be sub-optimal in onionshare's httpd, and this may still be a bottleneck after performance improvements are deployed for Tor.

Point 1 means that the optimal behavior for the current situation is to use two downloads per circuit, to fill the circuit window. To avoid circuit rebuild, SOCKS username and password should be re-used, with the same U+P reused for pairs of two concurrent range requests. Additional socks U+P can be used for additional pairs of requests, to create more circuits. Again, these U+P should be re-used for subsequent chunks, to avoid circuit rebuild.

It also suggests that perhaps http range chunks should be around 250000 bytes each, I think. But this should be investigated (500 cell windows of 500 bytes is 250000 bytes).

If this is done properly, in theory, onionshare downloads should be able to hit 6 megabytes/sec if fast guards are in use and a circuit build timeout is learned, even before we deploy any performance improvements in Tor clients.

Point 5 suggests that your httpd may also be a bottleneck. So absent even deploying this feature, looking into your httpd more closely will benefit you long term.

Also, you probably should not overdo this feature with an extremely large number of circuits, especially after congestion control is deployed, as this may begin to cause imbalance and unfairness in the network.

Thank you for all your work, Micah!

**Comment by mikeperry-tor** [comment]
One more thing: you can track these http range requests, to ensure all chunks complete, to preserve the feature that onionshare shuts down after the file is complete. This will now need to happen at the httpd layer, if it was previously happening elsewhere.

**Comment by micahflee** [comment]
Wow, 6 megabytes/sec sounds AMAZING. I definitely want to make this happen.

**Comment by mikeperry-tor** [comment]
> Wow, 6 megabytes/sec sounds AMAZING. I definitely want to make this happen.

Yes, and note that 6 megabytes/sec was the limit we hit with ONE guard on ONE six-hop circuit. This limit exists because of our KIST socket scheduler, which limits how fast relays and clients can read/write on a single TLS connection. Tor should be switching to two guards soon, for an overwhelmingly large number of reasons. Once that happens, 12 megabytes/sec will become your new limit (because you'll be making circuits across both guards).

And, once we deploy congestion control, we will re-investigate tuning or removing KIST, to remove the 6 megabyte/sec TLS limit. After that, pure gravy.
